### PR TITLE
Add controllers

### PR DESCRIPTION
**Pull request by thumbnail**
- Removed original abstractions of agent and controllers
 - Add `RebindingPublisher`
 - Add `ControllerSetPublisher`
 - Add `ControllerHandler`

`ControllerIndex` is used instead of `ControllerState` since index is lost but used to do the mapping.

### CODE REVIEW

File: src/main/java/io/leonis/zosma/ipc/controllers/RemapFunction.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import io.leonis.zosma.game.data.Player.PlayerIdentity;
+import io.reactivex.functions.BiFunction;
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+/**
+ * @author jeroen.dejong.
+ */
+public final class RemapFunction implements
```
Comment (thumbnail) [inline src/main/java/io/leonis/zosma/ipc/controllers/RemapFunction.java]: Rename to `DpadRemapFunction` ? also javadoc is missing

File: src/main/java/io/leonis/zosma/ipc/controllers/ControllerSetPublisher.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * The Class ControllerSetPublisher.
+ *
+ * Publishes all <b>connected</b> controllers at a provided interval.
+ *
+ * @author Jeroen de Jong
+ */
+public final class ControllerSetPublisher implements Publisher<Set<ControllerState>> {
+
+  /**
+   * Interval between polls.
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/controllers/ControllerSetPublisher.java]: in ms

File: src/main/java/io/leonis/zosma/ipc/controllers/ControllerSetPublisher.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import io.reactivex.Flowable;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.*;
+import org.reactivestreams.*;
+
+/**
+ * The Class ControllerSetPublisher.
+ *
+ * Publishes all <b>connected</b> controllers at a provided interval.
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/controllers/ControllerSetPublisher.java]: why the `<b>`? if you really want to stress the importance i guess use `<strong>` but i think it's unnecessary

File: src/main/java/io/leonis/zosma/ipc/controllers/ControllerTransformer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+
+/**
+ * The Class ControllerTransformer.
+ *
+ * Transforms a {@link Set} of {@link ControllerIndex Controllers} to a {@link PlayerCommand}, and
+ * map it grouped by it's applicable identity.
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/controllers/ControllerTransformer.java]: what does `map it grouped` mean? also it's `its`

File: src/main/java/io/leonis/zosma/ipc/controllers/ReassignControllersFunction.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    || (controller.isButtonJustPressed(DPAD_RIGHT)
+                    && mapping.get(ids.lower(entry.getKey())).contains(controller.getIndex()))
+                    ) {
+                  return ImmutableSet.<Integer>builder()
+                      .addAll(entry.getValue())
+                      .add(controller.getIndex())
+                      .build();
+                }
+                // FIXME solve this try-catch thing
+              } catch (ControllerUnpluggedException e) {
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/controllers/ReassignControllersFunction.java]: rethrow this out of the lambda and add the exception to the signature of `apply`? it's a `io.reactivex.functions.BiFunction` so it shouldn't be a problem

File: src/main/java/io/leonis/zosma/ipc/controllers/RebindingPublisher.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  private final Flowable<Set<ControllerIndex>> controllerSetPublisher;
+  private final BiFunction<Map<PlayerIdentity, Set<Integer>>, ControllerIndex, Map<PlayerIdentity, Set<Integer>>> remapper;
+
+  @Override
+  public void subscribe(final Subscriber<? super Map<PlayerIdentity, Set<Integer>>> subscriber) {
+    this.controllerSetPublisher
+        .map(controllers ->
+            controllers.stream()
+                .reduce(this.originalMapping, this.remapper,
+                    (a, b) -> Stream.concat(a.entrySet().stream(), b.entrySet().stream())
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/controllers/RebindingPublisher.java]: `a` and `b` should probably have different names

**Review by romni — APPROVED** [review]

**Pull request merged**